### PR TITLE
Removing MediaRecorderErrorEvent, in favour of just a DOMException

### PR DESCRIPTION
**Pull request by yell0wd0g**
This is my first attempt at removing the MediaRecorderErrorEvent (https://github.com/w3c/mediacapture-record/issues/14). I included the list of potential DOMExceptions generated. 

I'm absolutely 0% sure I'm writing the right w3c lingo, please review.

### CODE REVIEW

File: MediaRecorder.html [hunk 1]
```diff
@@ -340,149 +340,87 @@
...
- </section>
-
-
+  <section id="error-handling">
+    <h2>Error Handling</h2>
 
+    <section id="general-principles">
+      <h3>General Principles</h3>
 
+      <p>Errors are indicated via <a href="http://www.w3.org/TR/2012/WD-dom-20121206/#exception-domexception">DOMException</a>s raised (see [[!DOM4]]). An exception <em title="must" class="rfc2119">must</em> be thrown when the error can be detected at the time that the call is made. If recording has been started and not yet stopped when the error occurs, then after raising the error, the UA <em title="must" class="rfc2119">must</em> raise a  dataavailable event, containing any data that it has gathered, and then a stop event. The UA  <em title="may" class="rfc2119">may</em> set platform-specific limits, such those for the minimum and maximum Blob size that it will support, or the number of Tracks it will record at once.  It <em title="must" class="rfc2119">must</em> signal a fatal error if these limits are exceeded. </p>
```
Comment (foolip) [inline MediaRecorder.html]: DOMException is defined by WebIDL, so https://heycam.github.io/webidl/#idl-DOMException is the right reference. DOM itself (https://dom.spec.whatwg.org) references WebIDL for this.

Comment (foolip) [inline MediaRecorder.html]: Given that these are general principles, I would make the section non-normative and not use any RFC 2119 keywords. If there's anything here that isn't already said in one of the relevant algorithms elsewhere in the spec, that of course needs fixing first.

File: MediaRecorder.html [hunk 2]
```diff
@@ -340,149 +340,87 @@
...
+          <tr>
+            <td><code>UnknownError</code></td>
+            <td>The operation failed for an unknown transient reason (e.g. out of memory). User agents SHOULD provide as much additional information as possible in the <code>message</code> attribute.</td>
+          </tr>
+          <tr>
+            <td><code>NotSupportedError</code></td>
+            <td>A modification to the <code>stream</code> has occurred that makes it impossible to continue recording. An example would be the addition of a Track while recording is occurring. User agents SHOULD provide as much additional information  as possible in the <code>message</code> attribute.</td>
+          </tr>
+          <tr>
+            <td><code>SecurityError</code></td>
```
Comment (foolip) [inline MediaRecorder.html]: Searching in the document (your branch opened in a browser), SecurityError is the only exception I see referenced outside of this section. This section is non-normative, so some other part of the spec needs to say when these exceptions are thrown.

**Comment by foolip** [comment]
Had a look at the changes and commented in a few places. High-level, I think the biggest issues around events and exceptions are these:
1. Events are fired synchronously inside various algorithms, like the pause event in the pause() method. This is probably not intended, and not how it's implemented in Blink. Should probably say "queue a task to ..." and define what the [task source](https://html.spec.whatwg.org/#generic-task-sources) for those tasks is, and the "DOM manipulation task source" would probably be fine. (Probably some of the methods need to return and run [in parallel](https://html.spec.whatwg.org/#in-parallel).)
2. DOMError has been [removed](https://dom.spec.whatwg.org/#dom-core-changes) from DOM, and it was not an Event interface. For async errors, a simple event with name "error" plus an error attribute, or a custom error event interface like MediaRecorderErrorEvent, seems necessary.
3. The verb "raise" is used for both exceptions and events. The right terminology is e.g. "[throw](https://heycam.github.io/webidl/#dfn-throw) an InvalidStateError exception" and "[fire an event](https://dom.spec.whatwg.org/#concept-event-fire) named dataavailable".

**Comment by foolip** [comment]
Who is the implementor for Firefox? Some assurance that they will adopt the changes would be good.

**Comment by yell0wd0g** [comment]
Oops sorry I was addressing the first batch of changes and only saw your latest comments after uploading the added patch.

FTR, I think FF [1] has no MediaRecorderErrorEvent enumeration, which was the first pain point, and the one this patch addresses.

Also, partially related, uses a customised DOMException instead, such as foolip commented/suggested. May I suggest we limit this changes to the Exceptions, and leave the Event cleanups to another bug+patch?

[1] https://github.com/mozilla/gecko-dev/blob/master/dom/webidl/RecordErrorEvent.webidl

**Comment by foolip** [comment]
I like small patches, if you want to restore MediaRecorderErrorEvent here and only deal with exceptions.

**Comment by yell0wd0g** [comment]
Actually let me close this pull request and try another one from scratch :) Why?

I got confused with Exceptions and Events and mixed them both, let me follow with one single Pull Request tackling foolip@s suggestion of just dealing with DOMExceptions first, and another one with an attempt to clarify the Error Events. The first patch would be updating to what Gecko and Blink implementations do de facto, while the second one might be a tad less easy.

**Pull request closed without merging**